solutions/usaco-669.mdx

```diff
@@ -267,76 +267,138 @@ print(last_dist, file=open('moocast.out', 'w'))
 <CPPSection>
 
 ```cpp
-#include <bits/stdc++.h>
-using namespace std;
-using ll = long long;
-
-const int MAX_N = 1e3;
-
-vector<int> x(MAX_N), y(MAX_N);
-vector<bool> visited(MAX_N);
-vector<vector<bool>> g(MAX_N, vector<bool>(MAX_N, false));
-int n;
-
-int dfs(int v) {
-	visited[v] = true;
+#include <iostream>
+#include <fstream>
+#include <vector>
 
-	int curr_cows = 0;
-	for (int i = 0; i < n; i++) {
-		if (visited[i] || !g[v][i]) {
-			continue;
-		}
-		curr_cows += dfs(i);
-	}
+using std::cout;
+using std::endl;
+using std::vector;
+using std::pair;
 
-	return curr_cows + 1;
+int dist_sq(const pair<int, int>& p1, const pair<int, int>& p2) {
+	int dx = p1.first - p2.first;
+	int dy = p1.second - p2.second;
+	return dx * dx + dy * dy;
 }
 
-bool ok(ll mid) {
-	for (int i = 0; i < n; i++) {
-		for (int j = 0; j < n; j++) {
-			g[i][j] = false;
+bool all_reachable(int power, const vector<pair<int, int>>& cows) {
+	int start = 0;
+	vector<int> frontier{start};
+	vector<bool> reached(cows.size());
+	reached[start] = true;
+	while (!frontier.empty()) {
+		int curr = frontier.back();
+		frontier.pop_back();
+		for (int c = 0; c < cows.size(); c++) {
+			if (!reached[c] && dist_sq(cows[curr], cows[c]) <= power) {
+				frontier.push_back(c);
+				reached[c] = true;
+			}
 		}
 	}
 
-	for (int i = 0; i < n; i++) {
-		for (int j = 0; j < n; j++) {
-			ll dist = (x[i] - x[j]) * (x[i] - x[j]) + (y[i] - y[j]) * (y[i] - y[j]);
-			if (dist <= mid) {
-				g[i][j] = true;
-			}
+	for (bool c : reached) {
+		if (!c) {
+			return false;
 		}
 	}
-
-	fill(visited.begin(), visited.end(), false);
-	return dfs(0) == n;
+	return true;
 }
 
 int main() {
-	freopen("moocast.in", "r", stdin);
-	freopen("moocast.out", "w", stdout);
-
-	cin >> n;
-	for (int i = 0; i < n; i++) {
-		 cin >> x[i] >> y[i];
+	std::ifstream read("moocast.in");
+	int cow_num;
+	read >> cow_num;
+	vector<pair<int, int>> cows(cow_num);
+	for (pair<int, int>& c : cows) {
+		read >> c.first >> c.second;
 	}
 
-	ll lo = 1;
-	ll hi = INT32_MAX;
-	ll sol = hi;
+	int lo = 0;
+	int hi = INT32_MAX / 2;  // should be large enough
```

Comment:
i think it's better if u just do `1e9` and write in the comment: "arbitrarily large integer" or smtg

cuz this number feels too specific and might get some ppl confused

Reply:
kk